### PR TITLE
[Merged by Bors] - feat(Data/List/Change): add forwards induction for List.Chain and List.Chain'

### PR DESCRIPTION
**Pull request by Command-Master**
Moves:
- List.Chain.induction -> List.Chain.backwards_induction
- List.Chain.induction_head -> List.Chain.backwards_induction_head

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Chain.lean [hunk 1]
```diff
@@ -318,11 +318,35 @@ theorem exists_chain_of_relationReflTransGen (h : Relation.ReflTransGen r a b) :
...
+    exact h.induction p _ carries initial
+
 /-- Given a chain from `a` to `b`, and a predicate true at `b`, if `r x y → p y → p x` then
 the predicate is true everywhere in the chain and at `a`.
 That is, we can propagate the predicate up the chain.
 -/
-theorem Chain.induction (p : α → Prop) (l : List α) (h : Chain r a l)
+theorem Chain.backwards_induction (p : α → Prop) (l : List α) (h : Chain r a l)
     (hb : getLast (a :: l) (cons_ne_nil _ _) = b) (carries : ∀ ⦃x y : α⦄, r x y → p y → p x)
     (final : p b) : ∀ i ∈ a :: l, p i := by
```
Comment (YaelDillies) [inline Mathlib/Data/List/Chain.lean:351]: Can you prove this from `Chain.induction` applied to `l.reverse`?

File: Mathlib/Data/List/Chain.lean [hunk 2]
```diff
@@ -376,6 +400,25 @@ theorem Chain'.cons_of_le [LinearOrder α] {a : α} {as m : List α}
...
+lemma Chain'.chain {α : Type*} {R : α → α → Prop} {l : List α} {v : α}
+    (hl : l.Chain' R) (hv : (lne : l ≠ []) → R v (l.head lne)) : l.Chain R v := by
+  rw [List.chain_iff_get]
+  constructor
+  · intro h
+    rw [List.get_mk_zero]
+    apply hv
+  · exact List.chain'_iff_get.mp hl
+
+lemma Chain'.apply_eq {α : Type*} {f : α → α} {l : List α}
```
Comment (YaelDillies) [inline Mathlib/Data/List/Chain.lean:412]:
```suggestion
lemma Chain'.iterate_of_apply_eq {α : Type*} {f : α → α} {l : List α}
```

File: Mathlib/Data/List/Chain.lean (hunk 1 above)
Comment (Command-Master) [reply Mathlib/Data/List/Chain.lean:351]: Done, although I'm not sure if it's better.

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Comment by Ruben-VandeVelde** [comment]
It seems like `Chain.induction_head` was removed without a deprecation; any reason for that?

**Pull request closed without merging**